ViewTransition::updatePseudoElementStyles repeatedly flushes layout.

mattwoodrow opened the pull request:
#### 77c5d13c77da630c7b7a13f2af704378ba8e9866
<pre>
ViewTransition::updatePseudoElementStyles repeatedly flushes layout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=282448">https://bugs.webkit.org/show_bug.cgi?id=282448</a>
&lt;<a href="https://rdar.apple.com/139082479">rdar://139082479</a>&gt;

Reviewed by Tim Nguyen.

This code copies computed style properties which depend on layout, but also
marks renderers as needing layout and does so in a loop.

Split it into two steps, one that reads the style and sets the resulting values
as ::view-transition-group, and another which mutates the
RenderViewTransitionCapture renderer.

It also makes the style update an implicit part of captureNewState, so that we
can avoid a second iteration during activateViewTransition. Previously
updatePseudoElementStyles was blocked on a style flush from updating the dynamic
style sheet, even though it didn&apos;t read any related style data.

This results in less invalidation (since we no longer unconditionally call
didChangeStyleSheetContents on the initial update), and exposes a few bugs where
we weren&apos;t invalidating the individual changes correctly.

One fix is to make sure setting the image on a RenderViewTransitionCapture
correctly requests a repaint and layer contents change.

Also fixes an existing bug (that got exposed and started causing WPT failures
due to this change) where capturing an element into a view transition didn&apos;t
notify ancestors that the background would no longer be painted and obscuration
status needs to be computed.

* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::updatePropertiesForRenderer):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::updatePseudoElementStyles):
(WebCore::ViewTransition::updatePseudoElementSizes):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
(WebCore::RenderViewTransitionCapture::setImage):

Canonical link: <a href="https://commits.webkit.org/286209@main">https://commits.webkit.org/286209@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aefdfeac4b06d4012148b719c39e382ff02c92c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75206 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/54645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28046 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/26462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77323 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63785 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2430 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/79665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/26462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78273 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/49196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64596 "Build is in progress. Recent messages:") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/46558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/22096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/24788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/22436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/81147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2536 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/81147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2686 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/64600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/81147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10505 "Build is in progress. Recent messages:") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11601 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2497 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2522 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -736,6 +758,7 @@ void ViewTransition::handleTransitionFrame()
     }
 
     updatePseudoElementStyles();
+    updatePseudoElementSizes();
```
Is this slightly more expensive than the previous iteration? Due to looping twice through names instead of once.

mattwoodrow replied:
I guess slightly, yes. The reduction in wasted work massively compensates for that though :)

nt1m commented on Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
```diff
@@ -46,6 +46,10 @@ RenderViewTransitionCapture::~RenderViewTransitionCapture() = default;
 void RenderViewTransitionCapture::setImage(RefPtr<ImageBuffer> oldImage)
 {
     m_oldImage = oldImage;
+    if (hasLayer())
+        layer()->contentChanged(ContentChangeType::Image);
+    if (parent())
+        repaint();
```
Is this related?

mattwoodrow replied:
It is! Updated the commit message to explain it better.

mattwoodrow committed:
ViewTransition::updatePseudoElementStyles repeatedly flushes layout.
https://bugs.webkit.org/show_bug.cgi?id=282448
<rdar://139082479>

Reviewed by Tim Nguyen.

This code copies computed style properties which depend on layout, but also
marks renderers as needing layout and does so in a loop.

Split it into two steps, one that reads the style and sets the resulting values
as ::view-transition-group, and another which mutates the
RenderViewTransitionCapture renderer.

It also makes the style update an implicit part of captureNewState, so that we
can avoid a second iteration during activateViewTransition. Previously
updatePseudoElementStyles was blocked on a style flush from updating the dynamic
style sheet, even though it didn't read any related style data.

This results in less invalidation (since we no longer unconditionally call
didChangeStyleSheetContents on the initial update), and exposes a few bugs where
we weren't invalidating the individual changes correctly.

One fix is to make sure setting the image on a RenderViewTransitionCapture
correctly requests a repaint and layer contents change.

Also fixes an existing bug (that got exposed and started causing WPT failures
due to this change) where capturing an element into a view transition didn't
notify ancestors that the background would no longer be painted and obscuration
status needs to be computed.

* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::updatePropertiesForRenderer):
(WebCore::ViewTransition::captureNewState):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::handleTransitionFrame):
(WebCore::ViewTransition::updatePseudoElementStyles):
(WebCore::ViewTransition::updatePseudoElementSizes):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
(WebCore::RenderViewTransitionCapture::setImage):

Canonical link: https://commits.webkit.org/286209@main
